### ISSUE TITLE
node.warn(), debug nodes and Debug panel do not support arrays with additional properties

### ISSUE THREAD

**Issue by tilleul** [node-red/node-red#2643]
### What are the steps to reproduce?
Deploy the following flow
`
[
    {
        "id": "620e17cb.b82408",
        "type": "inject",
        "z": "c4be9dd8.3b604",
        "name": "",
        "props": [
            {
                "p": "payload"
            },
            {
                "p": "topic",
                "vt": "str"
            }
        ],
        "repeat": "",
        "crontab": "",
        "once": false,
        "onceDelay": 0.1,
        "topic": "",
        "payload": "",
        "payloadType": "date",
        "x": 110,
        "y": 80,
        "wires": [
            [
                "c3253e14.637de"
            ]
        ]
    },
    {
        "id": "c3253e14.637de",
        "type": "function",
        "z": "c4be9dd8.3b604",
        "name": "",
        "func": "const regex1 = RegExp('(?<foo>foo*)', 'g');\nconst str1 = 'table football, foosball';\nlet array1;\n\nwhile ((array1 = regex1.exec(str1)) !== null) {\n  console.dir(array1);\n  node.warn(array1);\n}\n\nlet a = [\"item 0\", \"item 1\", 2];\n\na[\"myProperty\"] = {\"prop 0\" : \"prop 0\", \"prop 1\": \"prop 1\"};\n\nconsole.dir(a);\nnode.warn(a);\n\nmsg.payload = a;\n\nreturn msg;",
        "outputs": 1,
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "x": 280,
        "y": 80,
        "wires": [
            [
                "2bb007b6.a72c78"
            ]
        ]
    },
    {
        "id": "2bb007b6.a72c78",
        "type": "debug",
        "z": "c4be9dd8.3b604",
        "name": "",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "false",
        "statusVal": "",
        "statusType": "auto",
        "x": 480,
        "y": 80,
        "wires": []
    }
]
`

### What happens?
The main code is in the function node
`const regex1 = RegExp('(?<foo>foo*)', 'g');`
`const str1 = 'table football, foosball';`
`let array1;`

`while ((array1 = regex1.exec(str1)) !== null) {`
`  console.dir(array1);`
`  node.warn(array1);`
`}`

`let a = ["item 0", "item 1", 2];`

`a["myProperty"] = {"prop 0" : "prop 0", "prop 1": "prop 1"};`

`console.dir(a);`
`node.warn(a);`

The result of `exec()` using a RegExp is an array with additional properties (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec#Syntax). Those properties are displayed when you log the array into the console but not when you display it in the Debug Panel (here using `node.warn()`, but the above flows shows that the Debug node doesn't do it correctly either). 

This is confusing.

The same result occurs when you directly add a property to an array like in the second part of the code above.

The correct output is the one given by the console:
`
[ 'foo',
  'foo',
  index: 6,
  input: 'table football, foosball',
  groups: { foo: 'foo' } ]
`
`
[ 'foo',
  'foo',
  index: 16,
  input: 'table football, foosball',
  groups: { foo: 'foo' } ]
`
`
[ 'item 0',
  'item 1',
  2,
  myProperty: { 'prop 0': 'prop 0', 'prop 1': 'prop 1' } ]
`

### What do you expect to happen?
The Debug Panel should display arrays including their properties if any.

### Please tell us about your environment:

- [x] Node-RED version: 1.1.0
- [x] Node.js version: 10.11
- [x] npm version: 6.4.1
- [x] Platform/OS: Windows 10 64 bits
- [x] Browser: Google Chrome  v83.0.4103.116

**Comment by dceejay** [comment]
"this is confusing" - well yes. so is adding properties to arrays :-)
I suspect we could possibly do a better job with the .warn() - but we may not with the debug node - as arrays with extra properties are not serialisable, and we want to make the node to node comms pluggable. Those extra properties also get stripped by many other operations like most iterations, .map etc.

**Comment by tilleul** [comment]
I agree that adding properties to arrays is probably a bad habit but unfortunately that's how exec() work (and string.match() as well -- and these two functions are widely used when you deal with regex -- maybe there are others ?). In fact, I came across this "bug" because I was trying to debug some code in a function node that used exec() and I couldn't understand where the "groups" property had disappeared ... until I printed the arrays contents in the console :) 
Now, there's no hurry. Now that I'm aware of the limitation I'll work around it but at least now you guys know that somebody fell into that trap and that something should be done to at least warn other users :)
